awx/main/tasks.py

```diff
@@ -1077,6 +1077,15 @@ def status_handler(self, status_data, runner_config):
             self.instance = self.update_model(self.instance.pk, job_args=json.dumps(runner_config.command),
                                               job_cwd=runner_config.cwd, job_env=job_env)
 
+    def check_handler(self, config):
+        '''
+        IsolatedManager callback triggered by the repeated checks of the isolated node
+        '''
+        self.instance = self.update_model(self.instance.pk,
+                                          job_args=json.dumps(config['command']),
+                                          job_cwd=config['cwd'],
+                                          job_env=build_safe_env(config['env']))
```

Comment:
`                    credential.credential_type.inject_credential(` <-- takes in and may modify self.safe_envs. We should repeat the logic performed here https://github.com/ansible/awx/blob/3f73176ef2ae0d952b02e579974c60b807ec172b/awx/main/tasks.py#L1074

_Ideally_ we would alter the code so we can converge the code for isolated and non-isolated.

Reply:
Is that why we did it this way?  I had wondered about that.

However, do we want to still do it in this case?  Somebody had suggested to me that the env might wind up with some different values on isolated nodes, which is why I decided to serialize it into an artifact file for transport back to the controlling node.

Reply:
`which is why I decided to serialize it into an artifact file for transport back to the controlling node` that part is ideal and correct. But those artifacts may contain sensitive vars so we need to redact it.

Reply:
That's what the `build_safe_env` call there is doing, if I understand things correctly.

Reply:
Yes, but that is one piece of the puzzle. The `safe_env` dict returned by `build_safe_env` is added to by `credential.credential_type.inject_credential()`. So we can't just call `build_safe_env`. We need to use the dict that has been added to by `credential.credential_type.inject_credential()`

Reply:
:face_with_head_bandage:

Reply:
Ok, this comment had me worried for a minute, but ultimately these are not the same dictionaries.  What I see is that in `BaseTask.run`:

- `build_safe_env(env)` gets stored as `self.safe_env`
- `inject_credential` gets called for every credential, which updates `env` and `self.safe_env`
- `IsolatedManager.run()` dispatches this stuff including the unredacted contents of `env` out to the isolated node
- the unredacted env vars as seen on the isolated node are then serialized to the `config` artifact that I've added to ansible-runner
- `IsolatedManager.check()` captures the contents of this artifact and passes it back to `BaseTask.check_handler()`
- and then the `check_handler` callback does the redaction by calling `build_safe_env`

Reply:
Yep, so you just need to get a handle on self.safe_env to do the redaction, rather than calling build_safe_env in check_handler.

Reply:
I guess I'm just not following.  It seems like the artifact `env` will not necessarily fully match `self.safe_env` anymore by the time the isolated node is done with it, so it seems appropriate to sanitize fresh from the value in the artifact instead of copying the values in `self.safe_env`.  And it looks like the credentials injector uses this utility to do the sanitization in the first place, so this should be performing the exact same data manipulation with the same type of results when I use it in `check_handler`:

https://github.com/ansible/awx/blob/3f73176ef2ae0d952b02e579974c60b807ec172b/awx/main/models/credential/__init__.py#L684

Let's talk in person or on a call, though, to convince each other.

Reply:
Ok, I think this change will sufficiently address the situation.